### PR TITLE
Add more completion providers and parameters auto-loading

### PR DESCRIPTION
**Pull request by zunlihu**
Hi, I just add more features to terraform extension.

- Resource, data source types Auto-completion with recommendation.
- Resource/data required parameters auto-loading.
- Module source types auto-completion with recommendation.
- Module required parameters auto-loading.
- Module optional parameters auto-completion with description and default value.
- Hyperlink for  module source types.

Note: My completion items include all providers in www.terraform.io and I give them different priority according to statistical results.

### CODE REVIEW

**Comment by josephholsten** [comment]
@zunlihu this is a rather large set of changes. Any chance you could break this into smaller PRs, I can't honestly imagine having the time to review a `+36,854 −1,347` changeset all at once.

File: .gitignore [hunk 1]
```diff
@@ -4,7 +4,4 @@ node_modules/
 out/
 .vscode-test
 npm-debug.log
-hcl-hil/*.js
-.terraform
-*.tfplan
-*.tfplan.json
+
```
Comment (mauve) [inline .gitignore]: please dont remove these lines

File: UserGuide.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  3. `module` `source` parameter auto-completion support for [all providers](https://registry.terraform.io/) with required parameters auto-loading.
+  
+  4. `resource | data` default(required) parameters auto-completion.
+  
+  5. `resource | data | module | output | variable` property types auto completion support
+  
+### Others
+  1. Hyperlink for resource, data types and module sources, which is linked to Example Usage Web.
+
+## Install from *.vsix in Jenkins
```
Comment (mauve) [inline UserGuide.md]: please remove the mentioning of Jenkins, I dont use it, feel free to mention Travis if you want to. However Travis doesn't really work with artefacts so I am planning to azure devops for building and testing

File: UserGuide.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+<h4 align="center">Adds intelligent auto-completion features for <a href="https://www.terraform.io/">Terraform</a> files</h4>
+
+
+<!-- markdownlint-enable -->
+<!-- markdownlint-disable MD002 MD013 MD041 -->
+
+## Features at a glance
+### Auto-completion features
+  1. `Terraform` high level types `(resource, data, variable, output,  module ...)` auto completion support.
+   - ![](https://github.com/zunlihu/Terraform-Language-Server/blob/master/images/Keyword.PNG) ```resource``` and `data` without bucket
```
Comment (mauve) [inline UserGuide.md]: please add these images to this repository

File: package.json [hunk 4]
```diff
@@ -3,8 +3,8 @@
   "name": "terraform",
   "displayName": "Terraform",
   "description": "Syntax highlighting, linting, formatting, and validation for Hashicorp's Terraform",
-  "version": "1.1.2",
-  "publisher": "mauve",
+  "version": "1.2.2",
+  "publisher": "zunli",
```
Comment (mauve) [inline package.json]: please don't.

you never need to change this, not even for local testing

File: src/autocompletion/completion-provider.ts [hunk 5]
```diff
@@ -106,8 +106,17 @@ export class CompletionProvider implements vscode.CompletionItemProvider {
...
+    let lineTillCurrentPosition = lineText.substr(0, position.character);
+    // Module source completion
+    let possibleSources: any[] = this.lookupForModuleSource(lineTillCurrentPosition);
+    if (possibleSources.length > 0) {
+      return this.getModuleAutoCompletion(possibleSources, vscode.CompletionItemKind.Class);
+    }
+
     if (ast) {
-      let offset1 = document.offsetAt(position);
+      let offset1 = document.
```
Comment (mauve) [inline src/autocompletion/completion-provider.ts]: weird linebreak

File: src/autocompletion/completion-provider.ts [hunk 6]
```diff
@@ -6,15 +6,15 @@ import { AstPosition, getTokenAtPosition } from '../index/ast';
...
+import { IFieldDef, IModuleArgsDef, allProviders, moduleSources, terraformConfigAutoComplete } from './model';
 import { SectionCompletions } from './section-completions';
 import { IndexLocator } from "../index/index-locator";
 
 const resourceExp = new RegExp("(resource|data)\\s+(\")?(\\w+)(\")?\\s+(\")?([\\w\\-]+)(\")?\\s+({)");
 const terraformExp = new RegExp("(variable|output|module)\\s+(\")?([\\w\\-]+)(\")?\\s+({)");
 const nestedRegexes: RegExp[] = [/\w[A-Za-z0-9\-_]*(\s*){/, /\w[A-Za-z0-9\-_]*(\s*)=(\s*){/];
 const propertyExp = new RegExp("^([\\w_-]+)$");
-
+let count: number = 1
```
Comment (mauve) [inline src/autocompletion/completion-provider.ts]: feels weird, does this really need to be a global?

File: src/autocompletion/model.ts [hunk 7]
```diff
@@ -1,18 +1,26 @@
-const awsResources: ITerraformData = require("../data/terraform-provider-aws.json");
-const azureResources: ITerraformData = require("../data/terraform-provider-azurerm.json");
-const googleResources: ITerraformData = require("../data/terraform-provider-google.json");
+// const awsResources: ITerraformData = require("../data/terraform-provider-aws.json");
```
Comment (mauve) [inline src/autocompletion/model.ts]: don't we need them anymore? if so, delete the files and these 3 lines

File: src/autocompletion/model.ts [hunk 8]
```diff
@@ -1,18 +1,26 @@
...
+// const azureResources: ITerraformData = require("../data/terraform-provider-azurerm.json");
+// const googleResources: ITerraformData = require("../data/terraform-provider-google.json");
 export const terraformConfigAutoComplete: ITerraformConfigAutoComplete = require("../data/terraform-config.json");
 
 import * as _ from "lodash";
 
-export const allProviders: ITerraformData = _.merge({}, awsResources, azureResources, googleResources);
+// export const allProviders: ITerraformData = _.merge({}, awsResources, azureResources, googleResources);
```
Comment (mauve) [inline src/autocompletion/model.ts]: if the line isnt needed, please delete

File: src/autocompletion/section-completions.ts [hunk 9]
```diff
@@ -45,22 +45,12 @@ function outputCompletion(): vscode.CompletionItem {
 }
 
 function resourceCompletion(): vscode.CompletionItem {
-    let item = new vscode.CompletionItem("resource", vscode.CompletionItemKind.Interface);
-    let snippet =
-        'resource "${1:type}" "${2:name}" {\n' +
-        '  $0\n' +
-        '}\n';
-    item.insertText = new vscode.SnippetString(snippet);
+    let item = new vscode.CompletionItem("resource", vscode.CompletionItemKind.Keyword);
```
Comment (mauve) [inline src/autocompletion/section-completions.ts]: why this change?

File: src/autocompletion/section-completions.ts [hunk 10]
```diff
@@ -45,22 +45,12 @@ function outputCompletion(): vscode.CompletionItem {
...
 }
 
 function dataCompletion(): vscode.CompletionItem {
-    let item = new vscode.CompletionItem("data", vscode.CompletionItemKind.Interface);
-    let snippet =
-        'data "${1:type}" "${2:name}" {\n' +
-        '  $0\n' +
-        '}\n';
-    item.insertText = new vscode.SnippetString(snippet);
+    let item = new vscode.CompletionItem("data", vscode.CompletionItemKind.Keyword);
```
Comment (mauve) [inline src/autocompletion/section-completions.ts]: why this change?

**Comment by paultyng** [comment]
We just released [v2.0.0-rc.1](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.0.0-rc.1) of the extension. The main features include:

* Added syntax support for 0.12
* Added [terraform-ls](https://github.com/hashicorp/terraform-ls) usage by default (currently on 0.3.0, which offers basic provider code completion)

You can find additional information and specifics in the [release notes](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.0.0-rc.1) and [CHANGELOG](https://github.com/hashicorp/vscode-terraform/blob/master/CHANGELOG.md).

With this release we expect that many of the prior issues and PRs are no longer relevant or have been addressed, and are therefore being closed. If you feel the action taken on an issue or PR is in error, please comment as such and we can figure out the appropriate way to address it.

We plan to add the final 2.0.0 release to the marketplace soon, but are actively seeking your feedback now on the release candidates. You can download the .vsix from the releases page and [manually install it in VS Code](https://code.visualstudio.com/docs/editor/extension-gallery#_install-from-a-vsix) to try it out.
